enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,335 @@
+
+
+### User Experience In Detail
+
+
+#### Encrypt a Virtual Machine Image
+
+
+1. Prepare a image
+
+2. Create a Secret
+    ```yaml
+    apiVersion: v1
+    kind: Secret
+    metadata:
+        name: longhorn-crypto
+        namespace: longhorn-system
+    stringData:
+        CRYPTO_KEY_VALUE: "Some secret value here"
+        CRYPTO_KEY_PROVIDER: "secret"
+        CRYPTO_KEY_CIPHER: "aes-xts-plain64"
+        CRYPTO_KEY_HASH: "sha256"
+        CRYPTO_KEY_SIZE: "256"
+        CRYPTO_PBKDF: "argon2i"
+    ```
+
+3. Create Storage Class
+    ```yaml
+    kind: StorageClass
+    apiVersion: storage.k8s.io/v1
+    metadata:
+      name: harvester-longhorn-encryption
+    provisioner: driver.longhorn.io
+    allowVolumeExpansion: true
+    parameters:
+        numberOfReplicas: "2"
+        staleReplicaTimeout: "2880" # 48 hours in minutes
+        fromBackup: ""
+        migratable: "true"
+        encrypted: "true"
+        csi.storage.k8s.io/provisioner-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/provisioner-secret-namespace: "longhorn-system"
+        csi.storage.k8s.io/node-publish-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/node-publish-secret-namespace: "longhorn-system"
+        csi.storage.k8s.io/node-stage-secret-name: "longhorn-crypto"
+        csi.storage.k8s.io/node-stage-secret-namespace: "longhorn-system"
+        Create Encrypted Virtual Machine Image:
+    ```
+
+4. Create Encrypted Virtual Machine Image
+    ```yaml
+    apiVersion: harvesterhci.io/v1beta1
+    kind: VirtualMachineImage
+    metadata:
+        name: encrypted-alpine-extended
+        namespace: default
+    spec:
+        displayName: encrypted-alpine-extended-3.20.2-x86_64.iso
+        retry: 3
+        sourceType: clone
+        storageClassParameters:
+            migratable: "true"
+            numberOfReplicas: "1"
+            staleReplicaTimeout: "30"
+        sourceParameters:
+            secretName: "longhorn-crypto"
+            secretNamespace: "longhorn-system"
+            encryption: "encrypt"
+            sourceVirtualMachineImageName: "image-fnw6n" # not display name
+            sourceVirtualMachineImageNamespace: "default"
+    ```
+
+5. Now, it's just like a normal image you could choose in GUI.
+    ![](./20240805-encrypted-block-volume/01.png)
+
+#### Decrypt a Virtual Machine Image
+
+
+1. Prepare an encrypted virtual machine image.
+
+2. Create an unencrypted virtual machine image
+    ```yaml
+    apiVersion: harvesterhci.io/v1beta1
+    kind: VirtualMachineImage
+    metadata:
+        name: decrypt-alpine-extended
+        namespace: default
+    spec:
+        displayName: decrypt-alpine-extended-3.20.2-x86_64.iso
+        retry: 3
+        sourceType: clone
+        storageClassParameters:
+            migratable: "true"
+            numberOfReplicas: "3"
+            staleReplicaTimeout: "30"
+        sourceParameters:
+            secretName: "longhorn-crypto"
```

Comment:
Any plan to protect the related secrets from being removed accidently?

Reply:
Good idea, we should avoid this happening.